src/WebJobs.Script.Analyzers/WebJobsAttributeAnalyzer.cs

```diff
@@ -0,0 +1,141 @@
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Text;
+
+namespace Microsoft.Azure.Functions.Analyzers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class WebJobsAttributeAnalyzer : DiagnosticAnalyzer
+    {
+        // TODO: Scope this to per-project
+        JobHostMetadataProvider _tooling;
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get { return ImmutableArray.Create(DiagnosticDescriptors.IllegalFunctionName); } }
+
+    public static void VerifyWebJobsLoaded() 
+        {
+            var x = new JobHost(new OptionsWrapper<JobHostOptions>(new JobHostOptions()), null);
+        }
+
+        public override void Initialize(AnalysisContext context)
+        {
+            // https://stackoverflow.com/questions/62638455/analyzer-with-code-fix-project-template-is-broken
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+
+            VerifyWebJobsLoaded();
+
+            // Analyze method signatures.
+            context.RegisterSyntaxNodeAction(AnalyzeMethodDeclarationNode, SyntaxKind.MethodDeclaration);
+
+            // Hook compilation to get the assemblies' references and build the WebJob tooling interfaces.
+            context.RegisterCompilationStartAction(AnalyzeCompilation);
+        }
+
+        private void AnalyzeCompilation(CompilationStartAnalysisContext context)
+        {
+            var compilation = context.Compilation;
+
+            AssemblyCache.Instance.Build(compilation);
+            _tooling = AssemblyCache.Instance.Tooling;
+
+            // cast to PortableExecutableReference which has a file path
+            var x1 = compilation.References.OfType<PortableExecutableReference>().ToArray();
+            var webJobsPath = (from reference in x1
+                               where IsWebJobsSdk(reference)
+                               select reference.FilePath).SingleOrDefault();
+
+            if (webJobsPath == null)
+            {
+                return; // Not a WebJobs project.
+            }
+        }
+
+        private bool IsWebJobsSdk(PortableExecutableReference reference)
+        {
+            if (reference.FilePath.EndsWith("Microsoft.Azure.WebJobs.dll"))
+            {
+                return true;
+            }
+            return false;
+        }
+
+        // This is called extremely frequently
+        // Analyze the method signature to validate binding attributes + types on the parameters
+        private void AnalyzeMethodDeclarationNode(SyntaxNodeAnalysisContext context)
+        {
+            if (_tooling == null) // Not yet initialized
+            {
+                return;
+            }
+
+            var methodDecl = (MethodDeclarationSyntax)context.Node;
+            var methodName = methodDecl.Identifier.ValueText;
+
+            if (!HasFunctionNameAttribute(context, methodDecl))
```

Comment:
Do we need the if condition and return block? Calling `HasFunctionNameAttribute` is good enough correct? I also think we should rename that method (HasFunctionNameAttribute) to be more explicit that it checks function name and report.

Reply:
Addressed in later comment.